### PR TITLE
KIP-482: add support for flexible versions / tagged fields

### PR DESCRIPTION
**Pull request by gabriel-tincu**
- Add support for varints
- Add support for compact collections (byte array, string, array)
- Add support for new request and response headers, supporting flexible
versions
- Add list of supported flexible versions, plus extraction script
- Add List / Alter partition reassignments apis 

---
This change is [Reviewable](https://reviewable.io/reviews/dpkp/kafka-python/2151)

### CODE REVIEW

**Comment by gabriel-tincu** [comment]
The way the request / response headers are handled needs an update, as apparently it's possible to send requests that have an api version with flexible versions enabled with both flexible versions and without. So if someone ends up defining a new requests object but using the old collections, introducing the empty tag buffer will mess up the entire payload. The java code base uses the json files to generate the classes so maybe we should look into that as well?

**Comment by jeffwidman** [comment]
Is there a KIP associated with this change?

**Comment by gabriel-tincu** [comment]
> Is there a KIP associated with this change? I switched teams and only work infrequently with Kafka now, so haven't kept up with the latest news.

@jeffwidman 
New types as described by the protocol docs: https://kafka.apache.org/protocol.html#protocol_types
Tagged fields KIP: https://cwiki.apache.org/confluence/display/KAFKA/KIP-482%3A+The+Kafka+Protocol+should+Support+Optional+Tagged+Fields
Jira issue requiring version bump to use both tagged fields and compact data types for apis: https://issues.apache.org/jira/browse/KAFKA-9420

File: kafka/admin/client.py [hunk 1]
```diff
@@ -460,6 +460,28 @@ def create_topics(self, new_topics, timeout_ms=None, validate_only=False):
         # TODO raise exceptions if errors
         return self._send_request_to_controller(request)
 
+    def list_partition_reassignments(self, topics_partitions, timeout_ms=None):
```
Comment (dpkp) [inline kafka/admin/client.py:463]: Given the complexity, I might suggest splitting this into a PR that adds support for KIP-482 at the protocol level and another PR that adds the new admin APIs.

File: kafka/protocol/api.py [hunk 2]
```diff
@@ -20,6 +20,22 @@ def __init__(self, request, correlation_id=0, client_id='kafka-python'):
...
 
 
+class RequestHeaderV2(Struct):
+    # Flexible response / request headers end in field buffer
+    SCHEMA = Schema(
+        ('api_key', Int16),
+        ('api_version', Int16),
+        ('correlation_id', Int32),
+        ('client_id', String('utf-8')),
+        ('_tag_buffer', TaggedFields)
```
Comment (dpkp) [inline kafka/protocol/api.py:30]: Can we name this simply `tags` ?

File: kafka/protocol/parser.py [hunk 3]
```diff
@@ -181,3 +190,11 @@ def _reset_buffer(self):
         self._receiving = False
         self._header.seek(0)
         self._rbuffer = None
+
+    @staticmethod
+    def is_flexible_version_request(request):
```
Comment (dpkp) [inline kafka/protocol/parser.py:195]: I think we should move this logic to the Request class as it seems like it's primary function is to determine which version of request and response headers we should be using for a particular api+version.

File: kafka/protocol/types.py [hunk 4]
```diff
@@ -196,3 +196,156 @@ def repr(self, list_of_items):
...
+class TaggedFields(AbstractType):
+    @classmethod
+    def decode(cls, data):
+        num_fields = UnsignedVarInt32.decode(data)
+        ret = {}
+        if not num_fields:
+            return ret
+        prev_tag = -1
+        for i in range(num_fields):
+            tag = UnsignedVarInt32.decode(data)
```
Comment (dpkp) [inline kafka/protocol/types.py:296]: Is it expected that tag numbers are globally unique? Is the apache kafka project planning to publish a list of supported / known tags + numbers, and value serialization formats?

**Review by dpkp — COMMENTED** [review]
> The way the request / response headers are handled needs an update, as apparently it's possible to send requests that have an api version with flexible versions enabled with both flexible versions and without.

Can you explain this in more detail? Do you mean that it is possible to send a "flexible version" api request but not include any tags in the header (as I understand it this takes up 1 byte)? Or do you mean not even include the tag structure in the header (0 bytes)? I would assume that any "flexible version" request would require at least a 1-byte empty tag buffer and would generate a response that also has at least a 1-byte empty tag buffer?

File: kafka/protocol/api.py [hunk 5]
```diff
@@ -20,6 +20,22 @@ def __init__(self, request, correlation_id=0, client_id='kafka-python'):
...
+    )
+
+    def __init__(self, request, correlation_id=0, client_id='kafka-python'):
```
[the remark this answers is not in the document]
Comment (gabriel-tincu) [reply kafka/protocol/api.py]: I'm adding constructor support for the tags, but for now they will still be initialized to empty i guess

File: kafka/protocol/parser.py [hunk 6]
```diff
@@ -59,9 +60,14 @@ def send_request(self, request, correlation_id=None):
         log.debug('Sending request %s', request)
         if correlation_id is None:
             correlation_id = self._next_correlation_id()
-        header = RequestHeader(request,
-                               correlation_id=correlation_id,
-                               client_id=self._client_id)
+        if not self.is_flexible_version_request(request):
```
[the remark this answers is not in the document]
Comment (gabriel-tincu) [reply kafka/protocol/parser.py]: Moved this and the response header parsing to the request class

File: kafka/protocol/parser.py [hunk 7]
```diff
@@ -162,6 +168,9 @@ def _process_response(self, read_buffer):
                 'Correlation IDs do not match: sent %d, recv %d'
                 % (correlation_id, recv_correlation_id))
 
+        # Flexible response / request headers end in field buffer
+        if self.is_flexible_version_request(request):
```
[the remark this answers is not in the document]
Comment (gabriel-tincu) [reply kafka/protocol/parser.py]: Just took your hint and added a class property to indicate this

File: kafka/protocol/parser.py [hunk 8]
```diff
@@ -162,6 +168,9 @@ def _process_response(self, read_buffer):
...
+            _ = TaggedFields.decode(read_buffer)
```
[the remark this answers is not in the document]
Comment (gabriel-tincu) [reply kafka/protocol/parser.py:162]: Added ResponseHeader classes

File: kafka/admin/client.py (hunk 1 above)
Comment (gabriel-tincu) [reply kafka/admin/client.py]: removed, will add in another PR

File: kafka/protocol/api.py (hunk 2 above)
Comment (gabriel-tincu) [reply kafka/protocol/api.py:30]: renamed

File: kafka/protocol/parser.py (hunk 3 above)
Comment (gabriel-tincu) [reply kafka/protocol/parser.py]: done

File: kafka/protocol/types.py (hunk 4 above)
Comment (gabriel-tincu) [reply kafka/protocol/types.py:310]: The documentation leads me to believe that, at least initially, there will not be any structure to these tags, other than what the client implementation may wish to impart on them (their example of using a trace id comes to mind)

**Comment by gabriel-tincu** [comment]
@dpkp this is ready for another look, whenever you have the time

**Comment by dpkp** [comment]
Thanks a lot. I really like this design. We will probably have to convert the default for FLEXIBLE_VERSIONS from false -> true eventually, but I think we can wait for this feature to bake upstream a bit before we do that.

**Comment by tvoinarovskyi** [comment]
Wow, this is awesome work! Good job! 👍

**Pull request merged**